File: frontend/src/stores/relays.ts
import { ref } from 'vue'
import { defineStore } from 'pinia'
import {
  createRelay,
  deleteRelay,
  listRelays,
  reconnectRelay as requestRelayReconnect,
  updateRelay,
} from '../generated/admin-api'
import type {
  ManagedRelay,
  ManagedRelayPatchRequest,
  ManagedRelayRequest,
} from '../generated/admin-api'
import { expectData, withData } from '../api'
import {
  STANDARD_PAGE_SIZE_OPTIONS,
  useStoredPageSize,
} from '../table-pagination'

export const useRelaysStore = defineStore('relays', () => {
  const relays = ref<ManagedRelay[]>([])
  const loading = ref(false)
  const reconnectingRelayId = ref<string | null>(null)
  const first = ref(0)
  const rows = useStoredPageSize('relays', 10, STANDARD_PAGE_SIZE_OPTIONS)
  const total = ref(0)
  const connectedCount = ref(0)
  const enabledCount = ref(0)

  async function refresh(
    nextFirst = first.value,
    nextRows = rows.value,
  ): Promise<void> {
    first.value = nextFirst
    rows.value = nextRows
    loading.value = true
    try {
      const response = expectData(
        await listRelays<true>(
          withData({ query: { first: first.value, rows: rows.value } }),
        ),
      )
      relays.value = response.relays
      total.value = response.total
      first.value = response.first
      rows.value = response.rows
      connectedCount.value = response.connected_count
      enabledCount.value = response.enabled_count
      if (
        relays.value.length === 0 &&
        total.value > 0 &&
        first.value >= total.value
      ) {
        const previousFirst =
          Math.floor((total.value - 1) / rows.value) * rows.value
        await refresh(previousFirst, rows.value)
      }
    } finally {
      loading.value = false
    }
  }

  async function saveRelay(
    relayId: string | null,
    body: ManagedRelayRequest | ManagedRelayPatchRequest,
  ): Promise<ManagedRelay> {
    const saved = relayId
      ? expectData(
          await updateRelay<true>(
            withData({
              path: { relay_id: relayId },
              body: body as ManagedRelayPatchRequest,
            }),
          ),
        )
      : expectData(
          await createRelay<true>(
            withData({
              body: body as ManagedRelayRequest,
            }),
          ),
        )
    await refresh()
    return saved
  }

  async function removeRelay(relayId: string): Promise<void> {
    await deleteRelay<true>(withData({ path: { relay_id: relayId } }))
    await refresh()
  }

  async function reconnectRelay(relayId: string): Promise<ManagedRelay> {
    reconnectingRelayId.value = relayId
    try {
      const relay = expectData(
        await requestRelayReconnect<true>(
          withData({
            path: { relay_id: relayId },
          }),
        ),
      )
      await refresh()
      return relay
    } finally {
      reconnectingRelayId.value = null
    }
  }

  return {
    connectedCount,
    enabledCount,
    first,
    loading,
    reconnectRelay,
    reconnectingRelayId,
    refresh,
    relays,
    removeRelay,
    saveRelay,
    rows,
    total,
  }
})

File: src/worker_admin/handlers/relays.rs
use super::*;
use super::{
    relay_input::{resolve_create_relay_input, resolve_update_relay_input},
    relay_validation::map_relay_db_error,
};

pub(super) async fn list_relays(
    State(state): State<AdminState>,
    headers: HeaderMap,
    Query(query): Query<TablePageQuery>,
) -> Response {
    if let Err(response) = ensure_admin(&state, &headers).await {
        return response;
    }
    let first = query.first.unwrap_or(0).max(0);
    let rows = query.rows.unwrap_or(20).clamp(1, 200);
    let (total, enabled_count, relays) = match state
        .config_repository
        .list_managed_relays_page(first, rows)
        .await
    {
        Ok(page) => page,
        Err(err) => return internal(&state, err),
    };
    let runtime = state.managed_relay_statuses.read().await.clone();
    let connected_count = runtime.values().filter(|status| status.connected).count() as i64;
    let relays = relays
        .into_iter()
        .map(|relay| {
            let status = runtime.get(&relay.relay_id).cloned().unwrap_or_default();
            relay.attach_runtime(status)
        })
        .collect();
    Json(ManagedRelayListResponse {
        relays,
        total,
        first,
        rows,
        connected_count,
        enabled_count,
    })
    .into_response()
}

pub(super) async fn get_relay(
    State(state): State<AdminState>,
    headers: HeaderMap,
    Path(relay_id): Path<Uuid>,
) -> Response {
    if let Err(response) = ensure_admin(&state, &headers).await {
        return response;
    }
    let relay = match state.config_repository.get_managed_relay(relay_id).await {
        Ok(Some(relay)) => relay,
        Ok(None) => return error(StatusCode::NOT_FOUND, "not_found", "relay not found"),
        Err(err) => return internal(&state, err),
    };
    Json(relay.attach_runtime(state.managed_runtime_status_or_default(relay_id).await))
        .into_response()
}

pub(super) async fn create_relay(
    State(state): State<AdminState>,
    headers: HeaderMap,
    Json(body): Json<ManagedRelayRequest>,
) -> Response {
    if let Err(response) = ensure_admin(&state, &headers).await {
        return response;
    }
    if let Err(message) = body.validate_create() {
        return bad_request(&message);
    }
    let input = match resolve_create_relay_input(&state, body).await {
        Ok(input) => input,
        Err(response) => return *response,
    };
    let relay = match state.config_repository.create_managed_relay(input).await {
        Ok(relay) => relay,
        Err(err) => return map_relay_db_error(&state, err),
    };
    if let Err(err) = state.reconcile_relays().await {
        return internal(&state, err);
    }
    let relay_id = relay.relay_id;
    Json(relay.attach_runtime(state.managed_runtime_status_or_default(relay_id).await))
        .into_response()
}

pub(super) async fn update_relay(
    State(state): State<AdminState>,
    headers: HeaderMap,
    Path(relay_id): Path<Uuid>,
    Json(body): Json<ManagedRelayPatchRequest>,
) -> Response {
    if let Err(response) = ensure_admin(&state, &headers).await {
        return response;
    }
    let existing = match state.config_repository.get_managed_relay(relay_id).await {
        Ok(Some(relay)) => relay,
        Ok(None) => return error(StatusCode::NOT_FOUND, "not_found", "relay not found"),
        Err(err) => return internal(&state, err),
    };
    if let Err(message) = body.validate_patch(existing.tls_mode) {
        return bad_request(&message);
    }
    let input = match resolve_update_relay_input(&state, existing.clone(), body).await {
        Ok(input) => input,
        Err(response) => return *response,
    };
    let relay = match state
        .config_repository
        .update_managed_relay(relay_id, input)
        .await
    {
        Ok(Some(relay)) => relay,
        Ok(None) => return error(StatusCode::NOT_FOUND, "not_found", "relay not found"),
        Err(err) => return map_relay_db_error(&state, err),
    };
    if let Err(err) = state.reconcile_relays().await {
        return internal(&state, err);
    }
    Json(relay.attach_runtime(state.managed_runtime_status_or_default(relay_id).await))
        .into_response()
}

pub(super) async fn delete_relay(
    State(state): State<AdminState>,
    headers: HeaderMap,
    Path(relay_id): Path<Uuid>,
) -> Response {
    if let Err(response) = ensure_admin(&state, &headers).await {
        return response;
    }
    match state.config_repository.delete_managed_relay(relay_id).await {
        Ok(true) => {
            if let Err(err) = state.reconcile_relays().await {
                return internal(&state, err);
            }
            state.managed_relay_statuses.write().await.remove(&relay_id);
            StatusCode::NO_CONTENT.into_response()
        }
        Ok(false) => error(StatusCode::NOT_FOUND, "not_found", "relay not found"),
        Err(err) => internal(&state, err),
    }
}

pub(super) async fn reconnect_relay(
    State(state): State<AdminState>,
    headers: HeaderMap,
    Path(relay_id): Path<Uuid>,
) -> Response {
    if let Err(response) = ensure_admin(&state, &headers).await {
        return response;
    }
    let relay = match state.config_repository.get_managed_relay(relay_id).await {
        Ok(Some(relay)) => relay,
        Ok(None) => return error(StatusCode::NOT_FOUND, "not_found", "relay not found"),
        Err(err) => return internal(&state, err),
    };
    if !relay.enabled {
        return bad_request("relay is disabled");
    }
    if let Err(err) = state.reconnect_relay(relay_id).await {
        return internal(&state, err);
    }
    Json(relay.attach_runtime(state.managed_runtime_status_or_default(relay_id).await))
        .into_response()
}
